### PR TITLE
Add gitlab-push function

### PR DESCRIPTION
**Pull request by martindekov**
Adding gitlab-push function which can be added to webhook route
to handle push or system hooks from gitlab

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

## Description

Abstract of how the function work + private/internal repository support:

gitlab push

This is clearer version of PR #109 . Validation of customers is removed along with webhook validation since those are given to `gitlab-event` function in another PR and this is no longer the front-facing function. Also internal trust/signing of the payload is added when calling `gitlab-status`.

## How Has This Been Tested?

Have `gitlab.yml` file with configuration:

```
provider:
  name: faas
  gateway: http://127.0.0.1:8080

functions:
  gitlab-push:
    lang: go
    handler: ./gitlab-push
    image: martindekov/gitlab-push:0.2
    labels:
      openfaas-cloud: "1"
      role: openfaas-system
    environment:
      write_debug: true
      read_debug: true
    environment_file:
      - gateway_config.yml
    secrets:
      - payload-secret
```

It is pretty standard

## How are existing users impacted? What migration steps/scripts do we need?

GitLab support for PushEvents

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

### CODE REVIEW

File: gitlab-push/handler_test.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	for _, test := range tests {
+		t.Run(test.title, func(t *testing.T) {
+			privateRepo := formatPrivateRepo(test.visibilityLevel)
+			if privateRepo != test.expectedBool {
+				t.Errorf("Unexpected result wanted: `%v` got: `%v`", test.expectedBool, privateRepo)
+			}
+		})
+	}
+
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler_test.go]: Maybe do this in a follow-up?

File: gitlab-push/handler.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	xCloudSignature := os.Getenv("Http_X_Cloud_Signature")
+
+	err = hmac.Validate(req, xCloudSignature, payloadSecret)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler.go]: Not sure these 5-7 lines warrant going in the SDK package.

File: gitlab-push/handler.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+			},
+			RepositoryURL: gitlabPushEvent.GitLabProject.WebURL,
+		},
+		AfterCommitID: gitlabPushEvent.AfterCommitID,
+		Installation: sdk.PushEventInstallation{
+			ID: gitlabPushEvent.GitLabProject.ID,
+		},
+	}
+
+	pushEvent.SCM = SCM
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler.go]: Makes sense

File: gitlab-push/handler.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		return fmt.Sprintf("%s cannot handle event: %s", Source, event)
+	}
+
+	gitlabPushEvent := sdk.GitLabPushEvent{}
+	err := json.Unmarshal(req, &gitlabPushEvent)
+	if err != nil {
+		return err.Error()
+	}
+
+	privateRepo := formatPrivateRepo(gitlabPushEvent.GitLabProject.VisibilityLevel)
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler.go]: Shorter name works for me.

File: gitlab-push/handler.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		log.Fatal(validateErr)
+	}
+
+	if audit == nil {
+		audit = sdk.AuditLogger{}
+	}
+
+	event := os.Getenv("Http_X_Gitlab_Event")
+
+	if event != "System Hook" {
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler.go]: Good suggestion

File: gitlab-push/handler.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+// Handle a serverless request
+func Handle(req []byte) string {
+	validateErr := validateRequest(req)
+	if validateErr != nil {
+		log.Fatal(validateErr)
+	}
+
+	if audit == nil {
+		audit = sdk.AuditLogger{}
+	}
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-push/handler.go]: Race condition how?

File: git-tar/function/ops_test.go [hunk 7]
```diff
@@ -100,3 +100,19 @@ func Test_getRepositoryURL_whenGetTokenReturnsError_WhenRepositoryIsPrivate(t *t
...
+	token := "xzajncx16zxc42"
+
+	expectedCloneURL := "https://martindekov:xzajncx16zxc42@gitlab.example.io/martindekov/gitlab-playground.git"
+	cloneURL, _ := formatGitLabCloneURL(event, token)
+	t.Run("Properly formatted URL", func(t *testing.T) {
+		if cloneURL != expectedCloneURL {
+			t.Errorf("Expected URL: %s got: %s", expectedCloneURL, cloneURL)
+		}
+	})
+
```
[the remark this answers is not in the document]
Comment (alexellis) [reply git-tar/function/ops_test.go]: Edge cases are important, at the minimum a positive and negative test.

File: git-tar/function/ops.go [hunk 8]
```diff
@@ -271,16 +271,35 @@ func clone(pushEvent sdk.PushEvent) (string, error) {
...
-	at := &githubAuthToken{
-		appID:          os.Getenv("github_app_id"),
-		installationID: pushEvent.Installation.ID,
-		privateKeyPath: sdk.GetPrivateKeyPath(),
+	var cloneURL string
+
+	if pushEvent.SCM == "github" {
+
+		at := &githubAuthToken{
+			appID:          os.Getenv("github_app_id"),
```
[the remark this answers is not in the document]
Comment (alexellis) [reply git-tar/function/ops.go]: Better as a follow-up issue.

File: gitlab-push/handler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivanayov) [reply gitlab-push/handler.go]: If two functions invoke this handler at the very same time and audit is nil for both.

File: gitlab-push/handler.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+func readBool(key string) bool {
+	if val, exists := os.LookupEnv(key); exists {
+		return val == "true" || val == "1"
+	}
+	return false
+}
+
+func reportGitLabStatus(status *sdk.Status) {
```
[the remark this answers is not in the document]
Comment (ivanayov) [reply gitlab-push/handler.go]: This is old comment. I have left unfinished review on Friday

File: gitlab-push/handler_test.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+		expectedBool    bool
+	}{
+		{
+			title:           "This is visibility level Private, users with access can see it",
+			visibilityLevel: 00,
+			expectedBool:    true,
+		},
+		{
+			title:           "This is visibility level Internal, only logged in users have access",
+			visibilityLevel: 10,
```
[the remark this answers is not in the document]
Comment (ivanayov) [reply gitlab-push/handler_test.go]: Also an old comment. Bart referenced it in the handler already.

File: gitlab-push/handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-push/handler.go]: I will do this in a follow up. For invalid values they will be set to private .. for now

File: gitlab-push/handler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-push/handler.go]: done

File: gitlab.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  gitlab-push:
+    lang: go
+    handler: ./gitlab-push
+    image: martindekov/gitlab-push:0.2
```
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab.yml]: Nice ^ Done

File: stack.yml [hunk 12]
```diff
@@ -60,6 +60,7 @@ functions:
     secrets:
       - payload-secret
       - private-key
+#      - gitlab-read-token
```
[the remark this answers is not in the document]
Comment (martindekov) [reply stack.yml]: Now it is like
```
# Uncomment this for GitLab
#      - gitlab-api-token
```

File: git-tar/function/ops_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply git-tar/function/ops_test.go]: Added ^

File: git-tar/function/ops.go [hunk 13]
```diff
@@ -271,18 +271,39 @@ func clone(pushEvent sdk.PushEvent) (string, error) {
...
+				return "", fmt.Errorf("cannot read api token from GitLab: %s", tokenErr.Error())
+			}
+
+			cloneURL, err = formatGitLabCloneURL(pushEvent, tokenAPI)
+		} else {
+			cloneURL = pushEvent.Repository.CloneURL
+		}
 	}
 
+	fmt.Printf("\n\n\n Clone URL is: `%s` \n\n\n", cloneURL)
```
[the remark this answers is not in the document]
Comment (martindekov) [reply git-tar/function/ops.go]: Done

**Comment by alexellis** [comment]
Thanks @martindekov

**Comment by alexellis** [comment]
Please go ahead and raise a PR to master now from integrate/gitlab.

**Comment by martindekov** [comment]
On it
